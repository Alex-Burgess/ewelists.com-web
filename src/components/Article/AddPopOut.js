import React, { useState } from 'react';
import update from 'immutability-helper';
import { API } from "aws-amplify";
// libs
import { onError, debugError } from "libs/errorLib";
// nodejs library to set properties for components
import PropTypes from "prop-types";
// @material-ui/core components
import { makeStyles } from "@material-ui/core/styles";
import Dialog from "@material-ui/core/Dialog";
import DialogContent from "@material-ui/core/DialogContent";
import Slide from "@material-ui/core/Slide";
// @material-ui/icons
import Close from "@material-ui/icons/Close";
import Add from "@material-ui/icons/Add";
import Done from "@material-ui/icons/Done";
// core components
import Button from "components/Buttons/Button.js";
import DialogTitle from "@material-ui/core/DialogTitle";
import Table from "components/Table/Table.js";
import CircularProgress from '@material-ui/core/CircularProgress';

import styles from "assets/jss/material-kit-pro-react/components/article/addPopoutStyle.js";
const useStyles = makeStyles(styles);

const Transition = React.forwardRef(function Transition(props, ref) {
  return <Slide direction="down" ref={ref} {...props} />;
});

Transition.displayName = "Transition";

export default function AddPopOut(props) {
  const classes = useStyles();
  const { open, lists, product } = props;
  const [added, setAdded] = useState({});
  const [adding, setAdding] = useState({});
  const [addError, setAddError] = useState({});

  const closeEditPopOut = () => {
    clearError();
    props.handleClose(product['productId']);
  }

  const updateAdding = (state, id) => {
    if (state) {
      setAdding({
        ...adding,
          [id]: true
      })
    } else {
      setAdding({
        adding: update(adding, {
          [id]: {$set: false}
        })
      })
    }
  }

  const clearError = (id) => {
    if (id) {
      setAddError({
        addError: update(addError, {
          [id]: {$set: false}
        })
      })
    } else {
      setAddError({});
    }
  }

  const addProductToList = async (listId) => {
    updateAdding(true, listId);
    clearError(listId);

    let productId = product.productId;
    debugError("adding product (" + productId + ") to list: (" + listId + ")");

    try {
      await API.post("lists", "/" + listId + "/product/" +  productId, {
        body: {
          "quantity": "1",
          "productType": "products"
        }
      });
      debugError("Added product to list.")
    } catch (e) {
      onError(e);

      if (e.response.data.error !== 'Product already exists in list.') {
        setAddError({
          ...addError, [listId]: 'Could not be added.'
        })

        updateAdding(false, listId);
        return false
      }
    }

    setAdded({
      ...added, [listId]: true
    })
    updateAdding(false, listId);
    return true
  }

  const renderListsTable = () => {
    return (
      <Table
        tableData={
          renderRow()
        }
        tableShopping
        customCellClasses={[
          classes.listText,
          classes.buttonCell,
        ]}
        customClassesForCells={[0, 1]}
      />
    )
  }

  const renderButton = (id, list) => {
    if ((list.products.indexOf(product.productId) > -1 ) || (added[id])) {
      return (
        <Button round justIcon color="success" disabled className={classes.successButton}>
          <Done />
        </Button>
      )
    } else if(adding[id]) {
      return (
        <div className={classes.progressWrapper}>
          <Button round justIcon color="primary" disabled>
            <Add />
          </Button>
          <CircularProgress className={classes.fabProgress} size={51}/>
        </div>
      )
    } else {
      return (
        <Button round justIcon color="primary" onClick={() => addProductToList(id)}>
          <Add />
        </Button>
      )
    }
  }


  const renderRow = () => {
    if (Object.keys(lists).length > 0) {
      return Object.entries(lists).map(
        ([id, list]) =>
          [
            <span>
              <small className={classes.tdNameSmall}>
                {list.title}
                { addError[id]
                  ? <div className={classes.addError}>{addError[id]}</div>
                  : null
                }
              </small>
            </span>,
            <span>
              {renderButton(id, list)}
            </span>
          ]
      )
    } else {
      return [
        [
          <span>
            <small className={classes.tdNameSmall}>
              You have no lists yet...
            </small>
          </span>,
          <span>
            <a href={"/?create"} target="_blank" rel="noopener noreferrer">
              <Button round color="primary">
                Create a list
              </Button>
            </a>
          </span>
        ]
      ]
    }

  }

  return (
    <div className={classes.section}>
      {/* NOTICE MODAL START */}
      <Dialog
        classes={{
          root: classes.modalRoot,
          paper: classes.modal
        }}
        open={open}
        TransitionComponent={Transition}
        keepMounted
        onClose={() => closeEditPopOut()}
        aria-labelledby="notice-modal-slide-title"
        aria-describedby="notice-modal-slide-description"
      >
      <DialogTitle
        id="notice-modal-slide-title"
        disableTypography
        className={classes.modalHeader}
      >
        <Button
          simple
          className={classes.modalCloseButton}
          key="close"
          aria-label="Close"
          onClick={() => closeEditPopOut()}
        >
          {" "}
          <Close className={classes.modalClose} />
        </Button>
        <h4 className={classes.modalTitle}>
          Add Item to List
        </h4>
      </DialogTitle>
      <DialogContent
        id="notice-modal-slide-description"
        className={classes.modalBody}
      >
      <div className={classes.productImageContainer}>
        <img src={product['imageUrl']} className={classes.productImage} alt=".." />
      </div>
      <div>
        <h4 className={classes.cardTitle}>{product.brand}</h4>
      </div>
      <div>
        <p className={classes.description}>
          {product.description}
        </p>
      </div>
      {renderListsTable()}
      </DialogContent>
      </Dialog>
      {/* NOTICE MODAL END */}
    </div>
  );
}

AddPopOut.propTypes = {
  open: PropTypes.bool,
  lists: PropTypes.object,
  product: PropTypes.object
};
