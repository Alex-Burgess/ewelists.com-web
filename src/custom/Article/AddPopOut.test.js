import * as React from 'react';
import {mount, shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import AddPopOut from './AddPopOut';

const product = {
  "productId": "12345678-prod-h012-1234-abcdefghijkl",
  "brand": "John Lewis and Partners",
  "details": "Grobag Swaddle Blanket, Pack of 2, White",
  "imageUrl": "https://johnlewis.scene7.com/is/image/JohnLewis/236972855?$rsp-pdp-port-640$",
  "productUrl": "https://www.johnlewis.com/grobag-swaddle-blanket-pack-of-2-white/p3310657"
};

const lists = {"6036fe33-ac32-47aa-9b75-52e585fda530":{"title":"Test List","products":[]}};

const lists2 = {"6036fe33-ac32-47aa-9b75-52e585fda530":{"title":"Test List","products":["12345678-prod-h012-1234-abcdefghijkl"]}}

const lists3 = {
  "6036fe33-ac32-47aa-9b75-52e585fda530":{"title":"Test List","products":["12345678-prod-h012-1234-abcdefghijkl"]},
  "1234abcd-ac32-47aa-9b75-52e585fda531":{"title":"Test List2","products":["12345678-prod-h001-1234-abcdefghijkl"]}
}

// TODO - snapshots are not really that clear.
// TODO - should I use mount?
// TODO - should I be testing the output in more detail, e.g. find the table and check the number of rows.

describe('Snapshot tests for add pop out', () => {
  let component = shallow(
    <AddPopOut
      open={true}
      lists={{}}
      product={product}
    />
  );

  it('Renders pop out with no lists.', () => {

    expect(toJson(component)).toMatchSnapshot();
  });

  it('Renders pop out with one list in table.', () => {
    const component = shallow(
      <AddPopOut
        open={true}
        lists={lists}
        product={product}
      />
    );

    expect(toJson(component)).toMatchSnapshot();
  });

  it('Renders pop out with list showing already added.', () => {
    const component = shallow(
      <AddPopOut
        open={true}
        lists={lists2}
        product={product}
      />
    );

    expect(toJson(component)).toMatchSnapshot();
  });

  it('Renders pop out with multiple lists.', () => {
    const component = shallow(
      <AddPopOut
        open={true}
        lists={lists3}
        product={product}
      />
    );

    expect(toJson(component)).toMatchSnapshot();
  });
})
