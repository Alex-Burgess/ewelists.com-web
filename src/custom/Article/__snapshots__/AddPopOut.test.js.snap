// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests for add pop out Renders pop out with list showing already added. 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    TransitionComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Transition",
        "render": [Function],
      }
    }
    aria-describedby="notice-modal-slide-description"
    aria-labelledby="notice-modal-slide-title"
    classes={
      Object {
        "paper": "makeStyles-modal-2",
        "root": "makeStyles-modalRoot-1",
      }
    }
    keepMounted={true}
    onClose={[Function]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      className="makeStyles-modalHeader-3"
      disableTypography={true}
      id="notice-modal-slide-title"
    >
      <ForwardRef
        aria-label="Close"
        className="makeStyles-modalCloseButton-5"
        key="close"
        onClick={[Function]}
        simple={true}
      >
         
        <CloseIcon
          className="makeStyles-modalClose-6"
        />
      </ForwardRef>
      <h4
        className="makeStyles-modalTitle-4"
      >
        Add Item to List
      </h4>
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))
      className="makeStyles-modalBody-7"
      id="notice-modal-slide-description"
    >
      <div
        className="makeStyles-productImageContainer-22"
      >
        <img
          alt=".."
          className="makeStyles-productImage-21"
          src="https://johnlewis.scene7.com/is/image/JohnLewis/236972855?$rsp-pdp-port-640$"
        />
      </div>
      <div>
        <h4
          className="makeStyles-cardTitle-23"
        >
          John Lewis and Partners
        </h4>
      </div>
      <div>
        <p
          className="makeStyles-description-24"
        />
      </div>
      <CustomTable
        coloredColls={Array []}
        colorsColls={Array []}
        customCellClasses={
          Array [
            "makeStyles-listText-26",
            "makeStyles-buttonCell-30",
          ]
        }
        customClassesForCells={
          Array [
            0,
            1,
          ]
        }
        customHeadCellClasses={Array []}
        customHeadClassesForCells={Array []}
        hover={false}
        striped={false}
        tableData={
          Array [
            Array [
              <span>
                <small
                  className="makeStyles-tdNameSmall-25"
                >
                  Test List
                </small>
              </span>,
              <span>
                <ForwardRef
                  className="makeStyles-successButton-27"
                  color="success"
                  disabled={true}
                  justIcon={true}
                  round={true}
                >
                  <UNDEFINED />
                </ForwardRef>
              </span>,
            ],
          ]
        }
        tableHeaderColor="gray"
        tableShopping={true}
      />
    </WithStyles(ForwardRef(DialogContent))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;

exports[`Snapshot tests for add pop out Renders pop out with multiple lists. 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    TransitionComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Transition",
        "render": [Function],
      }
    }
    aria-describedby="notice-modal-slide-description"
    aria-labelledby="notice-modal-slide-title"
    classes={
      Object {
        "paper": "makeStyles-modal-2",
        "root": "makeStyles-modalRoot-1",
      }
    }
    keepMounted={true}
    onClose={[Function]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      className="makeStyles-modalHeader-3"
      disableTypography={true}
      id="notice-modal-slide-title"
    >
      <ForwardRef
        aria-label="Close"
        className="makeStyles-modalCloseButton-5"
        key="close"
        onClick={[Function]}
        simple={true}
      >
         
        <CloseIcon
          className="makeStyles-modalClose-6"
        />
      </ForwardRef>
      <h4
        className="makeStyles-modalTitle-4"
      >
        Add Item to List
      </h4>
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))
      className="makeStyles-modalBody-7"
      id="notice-modal-slide-description"
    >
      <div
        className="makeStyles-productImageContainer-22"
      >
        <img
          alt=".."
          className="makeStyles-productImage-21"
          src="https://johnlewis.scene7.com/is/image/JohnLewis/236972855?$rsp-pdp-port-640$"
        />
      </div>
      <div>
        <h4
          className="makeStyles-cardTitle-23"
        >
          John Lewis and Partners
        </h4>
      </div>
      <div>
        <p
          className="makeStyles-description-24"
        />
      </div>
      <CustomTable
        coloredColls={Array []}
        colorsColls={Array []}
        customCellClasses={
          Array [
            "makeStyles-listText-26",
            "makeStyles-buttonCell-30",
          ]
        }
        customClassesForCells={
          Array [
            0,
            1,
          ]
        }
        customHeadCellClasses={Array []}
        customHeadClassesForCells={Array []}
        hover={false}
        striped={false}
        tableData={
          Array [
            Array [
              <span>
                <small
                  className="makeStyles-tdNameSmall-25"
                >
                  Test List
                </small>
              </span>,
              <span>
                <ForwardRef
                  className="makeStyles-successButton-27"
                  color="success"
                  disabled={true}
                  justIcon={true}
                  round={true}
                >
                  <UNDEFINED />
                </ForwardRef>
              </span>,
            ],
            Array [
              <span>
                <small
                  className="makeStyles-tdNameSmall-25"
                >
                  Test List2
                </small>
              </span>,
              <span>
                <ForwardRef
                  color="primary"
                  justIcon={true}
                  onClick={[Function]}
                  round={true}
                >
                  <UNDEFINED />
                </ForwardRef>
              </span>,
            ],
          ]
        }
        tableHeaderColor="gray"
        tableShopping={true}
      />
    </WithStyles(ForwardRef(DialogContent))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;

exports[`Snapshot tests for add pop out Renders pop out with no lists. 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    TransitionComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Transition",
        "render": [Function],
      }
    }
    aria-describedby="notice-modal-slide-description"
    aria-labelledby="notice-modal-slide-title"
    classes={
      Object {
        "paper": "makeStyles-modal-2",
        "root": "makeStyles-modalRoot-1",
      }
    }
    keepMounted={true}
    onClose={[Function]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      className="makeStyles-modalHeader-3"
      disableTypography={true}
      id="notice-modal-slide-title"
    >
      <ForwardRef
        aria-label="Close"
        className="makeStyles-modalCloseButton-5"
        key="close"
        onClick={[Function]}
        simple={true}
      >
         
        <CloseIcon
          className="makeStyles-modalClose-6"
        />
      </ForwardRef>
      <h4
        className="makeStyles-modalTitle-4"
      >
        Add Item to List
      </h4>
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))
      className="makeStyles-modalBody-7"
      id="notice-modal-slide-description"
    >
      <div
        className="makeStyles-productImageContainer-22"
      >
        <img
          alt=".."
          className="makeStyles-productImage-21"
          src="https://johnlewis.scene7.com/is/image/JohnLewis/236972855?$rsp-pdp-port-640$"
        />
      </div>
      <div>
        <h4
          className="makeStyles-cardTitle-23"
        >
          John Lewis and Partners
        </h4>
      </div>
      <div>
        <p
          className="makeStyles-description-24"
        />
      </div>
      <CustomTable
        coloredColls={Array []}
        colorsColls={Array []}
        customCellClasses={
          Array [
            "makeStyles-listText-26",
            "makeStyles-buttonCell-30",
          ]
        }
        customClassesForCells={
          Array [
            0,
            1,
          ]
        }
        customHeadCellClasses={Array []}
        customHeadClassesForCells={Array []}
        hover={false}
        striped={false}
        tableData={
          Array [
            Array [
              <span>
                <small
                  className="makeStyles-tdNameSmall-25"
                >
                  You have no lists yet...
                </small>
              </span>,
              <span>
                <a
                  href="/?create"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  <ForwardRef
                    color="primary"
                    round={true}
                  >
                    Create a list
                  </ForwardRef>
                </a>
              </span>,
            ],
          ]
        }
        tableHeaderColor="gray"
        tableShopping={true}
      />
    </WithStyles(ForwardRef(DialogContent))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;

exports[`Snapshot tests for add pop out Renders pop out with one list in table. 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    TransitionComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Transition",
        "render": [Function],
      }
    }
    aria-describedby="notice-modal-slide-description"
    aria-labelledby="notice-modal-slide-title"
    classes={
      Object {
        "paper": "makeStyles-modal-2",
        "root": "makeStyles-modalRoot-1",
      }
    }
    keepMounted={true}
    onClose={[Function]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      className="makeStyles-modalHeader-3"
      disableTypography={true}
      id="notice-modal-slide-title"
    >
      <ForwardRef
        aria-label="Close"
        className="makeStyles-modalCloseButton-5"
        key="close"
        onClick={[Function]}
        simple={true}
      >
         
        <CloseIcon
          className="makeStyles-modalClose-6"
        />
      </ForwardRef>
      <h4
        className="makeStyles-modalTitle-4"
      >
        Add Item to List
      </h4>
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))
      className="makeStyles-modalBody-7"
      id="notice-modal-slide-description"
    >
      <div
        className="makeStyles-productImageContainer-22"
      >
        <img
          alt=".."
          className="makeStyles-productImage-21"
          src="https://johnlewis.scene7.com/is/image/JohnLewis/236972855?$rsp-pdp-port-640$"
        />
      </div>
      <div>
        <h4
          className="makeStyles-cardTitle-23"
        >
          John Lewis and Partners
        </h4>
      </div>
      <div>
        <p
          className="makeStyles-description-24"
        />
      </div>
      <CustomTable
        coloredColls={Array []}
        colorsColls={Array []}
        customCellClasses={
          Array [
            "makeStyles-listText-26",
            "makeStyles-buttonCell-30",
          ]
        }
        customClassesForCells={
          Array [
            0,
            1,
          ]
        }
        customHeadCellClasses={Array []}
        customHeadClassesForCells={Array []}
        hover={false}
        striped={false}
        tableData={
          Array [
            Array [
              <span>
                <small
                  className="makeStyles-tdNameSmall-25"
                >
                  Test List
                </small>
              </span>,
              <span>
                <ForwardRef
                  color="primary"
                  justIcon={true}
                  onClick={[Function]}
                  round={true}
                >
                  <UNDEFINED />
                </ForwardRef>
              </span>,
            ],
          ]
        }
        tableHeaderColor="gray"
        tableShopping={true}
      />
    </WithStyles(ForwardRef(DialogContent))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;
